Add statuscode check on push

m-lukas opened the pull request:
Fix #45

m-lukas committed:
Add statuscode check to push command

This command add a new flag '-p/--ping' which can be used to check
the status of the application after it was pushed to ensure that
the application is working properly.

m-lukas committed:
Add tests for marshaling of new models

This commit adds two tests to model_test.go to check whether the
JSON of the routes and domain curls is parsed properly.

HeavyWombat commented on internal/gonut/cmd/push.go:
```diff
@@ -131,6 +132,7 @@ func init() {
 
 	pushCmd.PersistentFlags().StringVarP(&deleteSetting, "delete", "d", "always", "Delete application after push: always, never, on-success")
 	pushCmd.PersistentFlags().StringVarP(&summarySetting, "summary", "s", "short", "Push summary detail level: quiet, short, full")
+	pushCmd.PersistentFlags().BoolVarP(&pingSetting, "ping", "p", false, "Ping application after push to check the responding statuscode")
```
I would like to hear @tareqmamari opinion on this one, too. My take is that I would opt for the ping test being the default and the flag gives the end user the possibility to skip the check. The flag would be called `--no-ping` then.

HeavyWombat commented on internal/gonut/cmd/push.go:
```diff
@@ -208,18 +210,39 @@ func runSampleAppPush(app sampleApp) error {
 		return fmt.Errorf("unsupported delete setting: %s", deleteSetting)
 	}
 
+	// If summary is set to full, enable pinging to display it in the summary.
```
I do not like that we mix logic of the summary with checks. To be honest, I would like to remove this implicit enablement.

HeavyWombat commented on internal/gonut/cmd/push.go:
```diff
@@ -208,18 +210,39 @@ func runSampleAppPush(app sampleApp) error {
 		return err
 	}
 
+	// If ping setting is enabled but the summary is not set to full,
```
This breaks with the logic that follows afterwards. For example if summary is `json`, then `gonut` would print a text first and JSON output later. The whole output would not be JSON then. Same applies to the YAML output style. The oneline or short output would be affected, too. I would opt for removing this new part here and move the details in the respective outputs, for example to add the status code results in the exported report data.

HeavyWombat commented on internal/gonut/cmd/push.go:
```diff
@@ -259,6 +282,12 @@ func runSampleAppPush(app sampleApp) error {
 		bufPrintf("     DimGray{_stack:_} DarkSeaGreen{%s}\n", report.Stack())
 		bufPrintf(" DimGray{_buildpack:_} DarkSeaGreen{%s}\n", report.Buildpack())
+		// set color according to the statuscode
+		if report.StatusCode == 200 {
```
Just a minor thing: It helps with the readability to use `http.StatusOK` instead of `200`.

HeavyWombat commented on internal/gonut/cf/cf.go:
```diff
@@ -289,6 +304,27 @@ func getApp(appName string) (*AppDetails, error) {
+// getAppStatusCode sends a GET request to the given URL
+// and returns the statuscode.
+func getAppStatusCode(appRoute string) (int, error) {
+	resp, err := http.Get(appRoute)
```
One thing to investigate a little: Even though we do not use the response body, do we still need to add a `defer resp.Body.Close()` to make sure we close everything?

HeavyWombat commented on internal/gonut/cf/cf.go:
```diff
@@ -289,6 +304,27 @@ func getApp(appName string) (*AppDetails, error) {
+
+// getAppStatusCode sends a GET request to the given URL
+// and returns the statuscode.
+func getAppStatusCode(appRoute string) (int, error) {
```
I would like to suggest one more thing here: It is not just the HTTP response code that is interesting, it would also be interesting to notify the whole response time. Could you add a third return value, which is the overall time it took to get a result?

HeavyWombat reviewed (changes requested):
Thanks for the solid PR! I really like it. I noted a couple of changes I would like to see before we can merge it. We can also talk about it tomorrow.

m-lukas committed:
Reverse --ping flag logic

This commit introduces a new --no-ping flag which replaces the --ping flag.
Pinging will be executed by default.

m-lukas committed:
Close response body after pinging
